modules/core/04-channel/keeper/upgrade.go

```diff
@@ -95,6 +96,188 @@ func (k Keeper) WriteUpgradeInitChannel(
+		}
+
+		// this is first message in upgrade handshake on this chain so we must store original channel in restore channel path
+		// in case we need to restore channel later.
+		k.SetUpgradeRestoreChannel(ctx, portID, channelID, channel)
+	case types.INITUPGRADE:
+		upgradeSequence, found = k.GetUpgradeSequence(ctx, portID, channelID)
+		if !found {
+			errorReceipt := types.NewErrorReceipt(upgradeSequence, errorsmod.Wrapf(types.ErrInvalidUpgradeSequence, "upgrade sequence %d was not smaller than the counter party chain upgrade sequence %d", upgradeSequence, counterpartyUpgradeSequence))
+			upgradeSequence++
+
+			k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+			k.SetUpgradeSequence(ctx, portID, channelID, upgradeSequence)
+
+			return 0, "", errorsmod.Wrapf(types.ErrUpgradeAborted, "upgrade aborted, error receipt written for upgrade sequence: %d", upgradeSequence)
+		}
+
+		if upgradeSequence != counterpartyUpgradeSequence {
+			errorReceipt := types.NewErrorReceipt(upgradeSequence, errorsmod.Wrapf(types.ErrInvalidUpgradeSequence, "upgrade sequence %d was not smaller than the counter party chain upgrade sequence %d", upgradeSequence, counterpartyUpgradeSequence))
+			upgradeSequence++
+
+			k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+			k.SetUpgradeSequence(ctx, portID, channelID, upgradeSequence)
+			return 0, "", errorsmod.Wrapf(types.ErrUpgradeAborted, "upgrade aborted, error receipt written for upgrade sequence: %d", upgradeSequence)
+		}
+
+		// if there is a crossing hello, i.e an UpgradeInit has been called on both channelEnds,
+		// then we must ensure that the proposedUpgrade by the counterparty is the same as the currentChannel
+		// except for the channel state (upgrade channel will be in TRYUPGRADE and current channel will be in INITUPGRADE)
+		// if the proposed upgrades on either side are incompatible, then we will restore the channel and cancel the upgrade.
+		channel.State = types.TRYUPGRADE
+		k.SetChannel(ctx, portID, channelID, channel)
+
+		if !reflect.DeepEqual(channel, proposedUpgradeChannel) {
+			// TODO: log and emit events
+			if err := k.RestoreChannel(ctx, portID, channelID, upgradeSequence, types.ErrInvalidChannel); err != nil {
+				return 0, "", errorsmod.Wrap(types.ErrUpgradeAborted, err.Error())
+			}
+			return 0, "", errorsmod.Wrap(types.ErrUpgradeAborted, "proposed upgrade channel did not equal expected channel")
+		}
+
+		// retrieve restore channel to return previous version
+		restoreChannel, found := k.GetUpgradeRestoreChannel(ctx, portID, channelID)
+		if !found {
+			// TODO(this case should be unreachable): write error receipt for upgrade sequence and abort / cancel upgrade
+			return 0, "", errorsmod.Wrapf(types.ErrChannelNotFound, "upgrade aborted, error receipt written for upgrade sequence: %d", upgradeSequence)
+		}
```

Comment:
since this is technically unreachable should we panic? if it is reached then something is horribly wrong and some manual intervention or migration code is probably required.